layers/slices.go

```diff
@@ -0,0 +1,244 @@
+}
+
+func newSlicableDir(appDir string) (*sliceableDir, error) {
+	sdir := &sliceableDir{
+		path:        appDir,
+		slicedFiles: map[string]bool{},
+		pathInfos:   map[string]os.FileInfo{},
+		subDirs:     map[string][]string{},
+	}
+	if err := filepath.Walk(appDir, func(path string, fi os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		sdir.slicedFiles[path] = false
+		sdir.pathInfos[path] = fi
+		if fi.IsDir() {
+			children, err := ioutil.ReadDir(path)
+			if err != nil {
+				return err
+			}
+			for _, child := range children {
+				sdir.subDirs[path] = append(sdir.subDirs[path], filepath.Join(path, child.Name()))
+			}
+		}
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+	parentDirs, err := parents(appDir)
+	if err != nil {
+		return nil, err
+	}
+	sdir.parentDirs = parentDirs
+	return sdir, nil
+}
+
+func (sd *sliceableDir) sliceFiles(paths []string) []archive.PathInfo {
+	slicedFiles := map[string]os.FileInfo{}
+	for _, match := range paths {
+		sd.addMatchedFiles(slicedFiles, match)
+	}
+	return sd.fillInMissingParents(slicedFiles)
+}
+
+func (sd *sliceableDir) addMatchedFiles(matchedFiles map[string]os.FileInfo, match string) {
+	if added, ok := sd.slicedFiles[match]; !ok || added {
+		// don't add files that live outside the app dir
+		// don't add files were already added
+		return
+	}
+	if children, ok := sd.subDirs[match]; ok {
+		for _, child := range children {
+			sd.addMatchedFiles(matchedFiles, child)
+		}
+	}
+	matchedFiles[match] = sd.pathInfos[match]
+	sd.slicedFiles[match] = true
+}
+
+func (sd *sliceableDir) fillInMissingParents(matchedFiles map[string]os.FileInfo) []archive.PathInfo {
+	// add parent dirs for matched files if they are missing
+	var parentToCheck []string
+	var files []archive.PathInfo
+	addedParents := map[string]struct{}{}
+	for path, info := range matchedFiles {
+		files = append(files, archive.PathInfo{
+			Path: path,
+			Info: info,
+		})
+		parents := sd.fileParents(path)
+		for _, parent := range parents {
+			if _, ok := matchedFiles[parent.Path]; ok {
+				// don't add if it was slices as part of a match
+				continue
+			}
+			if _, ok := addedParents[parent.Path]; ok {
+				// don't add if it was already added
+				continue
+			}
+			parentToCheck = append(parentToCheck, parent.Path)
+			addedParents[parent.Path] = struct{}{}
+			files = append(files, parent)
+		}
+	}
+
+	// sort the dirs by their path depth (deepest -> most shallow) so we always process children first
+	sort.SliceStable(parentToCheck, func(i, j int) bool {
+		return len(strings.Split(parentToCheck[i], string(os.PathSeparator))) > len(strings.Split(parentToCheck[j], string(os.PathSeparator)))
+	})
+
+	// if all children are slices, mark the dir as sliced
+	for _, dir := range parentToCheck {
+		allChildrenAdded := true
+		if children, ok := sd.subDirs[dir]; ok {
+			for _, child := range children {
+				if added, ok := sd.slicedFiles[child]; ok && !added {
+					allChildrenAdded = false
```

Comment:
Should we stop iterating files at this point?

Reply:
yes, good catch